File: java/TwoPointer/Rearrange_Array_Elements_by_Sign_2149.java
package java.TwoPointer;

import java.util.*;
import java.io.*;

// import static java.lang.Math.max;
// import static java.lang.Math.min;
// import static java.lang.Math.abs;
// import static java.lang.System.out;

public class Rearrange_Array_Elements_by_Sign_2149 {

    public int[] rearrangeArray(int[] nums) {
        int i = 0, j = 0, k = 0, n = nums.length;
        int[] ans = new int[n];
        while (i < n || j < n) {
            while (i < n && nums[i] < 0)
                i++;
            if (i < n)
                ans[k++] = nums[i];
            while (j < n && nums[j] > 0)
                j++;
            if (j < n)
                ans[k++] = nums[j];
            ++i;
            ++j;
        }

        return ans;
    }

    public static void main(String[] args) throws FileNotFoundException {
        File f = new File("output.txt");

        // Scanner ip = new Scanner(System.in);
        try (Scanner ip = new Scanner(new File("input.txt"));
                PrintWriter op = new PrintWriter(f);) {
            Rearrange_Array_Elements_by_Sign_2149 obj = new Rearrange_Array_Elements_by_Sign_2149();
            int grid[] = { 3, 1, -2, -5, 2, -4 };
            op.println(obj.rearrangeArray(grid));

        }
    }

}

File: Code/TwoPointer/Rearrange_Array_Elements_by_Sign_2149.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> rearrangeArray(vector<int> nums)
    {
        vector<int> ans;
        int i = 0, j = 0, n = nums.size();
        while (i < n || j < n)
        {
            while (i < n && nums[i] < 0)
                i++;
            if (i < n)
                ans.push_back(nums[i]);
            while (j < n && nums[j] > 0)
                j++;
            if (j < n)
                ans.push_back(nums[j]);
            ++i, ++j;
        }

        return ans;
    }
};
int main()
{
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    Solution obj;
    auto ans = obj.rearrangeArray({3, 1, -2, -5, 2, -4});
    // cout << ans;
    return 0;
}